### PR TITLE
[scripts-audit] vcpkg_find_acquire_program

### PR DESCRIPTION
**Pull request by strega-nil-ms**
This is a PR for the actual changes to vcpkg_find_acquire_program, rather than trying to completely change how it works

### CODE REVIEW

**Comment by BillyONeal** [comment]
qca regressions were fixed by https://github.com/microsoft/vcpkg/pull/21250

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 1]
```diff
@@ -47,560 +47,634 @@ Note that msys2 has a dedicated helper function: [`vcpkg_acquire_msys`](vcpkg_ac
...
+                set(host_arch "$ENV{PROCESSOR_ARCHITECTURE}")
+            endif()
 
-  macro(do_find)
-    if(NOT DEFINED REQUIRED_INTERPRETER)
-      find_program(${VAR} ${PROGNAME} PATHS ${PATHS} NO_DEFAULT_PATH)
-      if(NOT ${VAR})
-        find_program(${VAR} ${PROGNAME})
-        if(${VAR} AND NOT ${VAR}_VERSION_CHECKED)
-            do_version_check()
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:495]: Here, there was a version check when the program was found from `ENV{PATH}`...

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 2]
```diff
@@ -47,560 +47,634 @@ Note that msys2 has a dedicated helper function: [`vcpkg_acquire_msys`](vcpkg_ac
...
-      file(MAKE_DIRECTORY ${PROG_PATH_SUBDIR})
-      if(DEFINED NOEXTRACT)
-        if(DEFINED _vfa_RENAME)
-          file(INSTALL ${ARCHIVE_PATH} DESTINATION ${PROG_PATH_SUBDIR} RENAME ${_vfa_RENAME} FILE_PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
-        else()
-          file(COPY ${ARCHIVE_PATH} DESTINATION ${PROG_PATH_SUBDIR} FILE_PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
+    z_vcpkg_find_acquire_program_find_internal("${program}"
+        INTERPRETER "${interpreter}"
+        PATHS ${paths_to_search}
+        NAMES ${search_names}
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:576]: ... but here, programs from apt or brew are no longer checked for version.

**Review by dg0yt — COMMENTED** [review]
I guess this broke building fontconfig.
https://github.com/microsoft/vcpkg/pull/21026#issuecomment-966438188

**Comment by BillyONeal** [comment]
Quoth @cenit : https://github.com/microsoft/vcpkg/pull/21026#issuecomment-966338216

> after this PR fontconfig has been broken on osx/linux in all PR
> 
> #20658 #20185

>I guess this broke building fontconfig.

Any ideas why we got a passing run here then?

**Comment by cenit** [comment]
my bad, my bisect was wrong

**Comment by Neumann-A** [comment]
> Any ideas why we got a passing run here then?

missing dep on tool-meson and the tool was probably downloaded and thus cached

```
Starting package 171/886: tool-meson:x64-linux
Building package tool-meson[core]:x64-linux...
-- Found meson('0.53.2') but at least version 0.58.1 is required! Trying to use internal version if possible!
-- Downloading https://github.com/mesonbuild/meson/archive/aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz -> meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz...
-- Using meson: /usr/bin/python3;/usr/bin/meson
-- Performing post-build validation
-- Performing post-build validation done
```


` Using meson: /usr/bin/python3;/usr/bin/meson` is strange if the internal stuff gets printed

**Comment by dg0yt** [comment]
> Any ideas why we got a passing run here the

Not exactly. But at least before this PR, vfap was not idempotent. (IIRC you never got the system version if a downloaded version was present.)

I tested fontconfig:x64-linux on a older state, before and after removing meson from `downloads/tools`. It does detect that a download is needed.
~~~
-- Found meson('') but at least version 0.58.1 is required! Trying to use internal version if possible!
-- Using cached meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz.
~~~

Will repeat with `master` now.

**Comment by dg0yt** [comment]
Well, I even didn't have meson installed until now. Now the results are even more interesting...

*On first run:*

~~~
-- Found meson('0.45.1') but at least version 0.58.1 is required! Trying to use internal version if possible!
-- Using cached meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz.
-- Configuring x64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: /usr/bin/python3 /usr/bin/meson --buildtype plain --backend ninja --wrap-mode nodownload --native /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/fontconfig/meson-native-x64-linux.log --default-library static --libdir lib --native /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/fontconfig/meson-native-x64-linux-debug.log -Ddebug=true --prefix /home/dg0yt/Projekte/vcpkg/vcpkg/packages/fontconfig_x64-linux/debug --includedir ../include -Dcmake_prefix_path=['/home/dg0yt/Projekte/vcpkg/vcpkg/installed/x64-linux/debug','/home/dg0yt/Projekte/vcpkg/vcpkg/installed/x64-linux'] /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/fontconfig/src/2.13.94-6fd83e1cbe.clean
    Working Directory: /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/fontconfig/x64-linux-dbg
    Error code: 2
    See logs for more information:
      /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/fontconfig/config-x64-linux-dbg-err.log
~~~

*On second run*, it just builds without meson messages and without error!
So still not idempotent, but in an unexpected way. 
**It downloads the newer tool in the first run, but use the outdate system version.** 
Now imagine some other port building with meson before fontconfig, but actually happy with the outdate version. CI passed.

(And I only noticed that because I canceled the first run when I noticed the missing system meson...)

**Comment by Neumann-A** [comment]
Ah it is probably not unsetting the variable after the first search.

**Comment by dg0yt** [comment]
On the positive side, it is not a regression :-) Now with meson installed, I see the same behaviour in the old tree.

**Comment by dg0yt** [comment]
> Ah it is probably not unsetting the variable after the first search.

More specific, I guess it is related to `${program}` being used both as a cache variable (via `find_program`) and as a normal variable. Unsetting the normal variable unhides the cached value.

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 3]
```diff
@@ -47,560 +47,634 @@ Note that msys2 has a dedicated helper function: [`vcpkg_acquire_msys`](vcpkg_ac
...
-      vcpkg_find_acquire_program(${REQUIRED_INTERPRETER})
-      find_file(SCRIPT_${VAR} NAMES ${SCRIPTNAME} PATHS ${PATHS} NO_DEFAULT_PATH)
-      if(NOT SCRIPT_${VAR})
-        find_file(SCRIPT_${VAR} NAMES ${SCRIPTNAME})
-        if(SCRIPT_${VAR} AND NOT ${VAR}_VERSION_CHECKED)
-            set(${VAR} ${${REQUIRED_INTERPRETER}} ${SCRIPT_${VAR}})
-            do_version_check()
-            set(${VAR}_VERSION_CHECKED ON)
-            if(NOT ${VAR})
-                unset(SCRIPT_${VAR} CACHE)
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:511]: notice: here the script var was unset in the cache not locally.

**Comment by Neumann-A** [comment]
> On the positive side, it is not a regression :-) Now with meson installed, I see the same behaviour in the old tree.

CI disagrees (the current main run just before these changes were merged). 
```
Starting package 214/1705: tool-meson:x64-linux
Building package tool-meson[core]:x64-linux...
-- Found external python3('Python 3.8.10').
-- Found meson('0.53.2') but at least version 0.58.1 is required! Trying to use internal version if possible!
-- Downloading https://github.com/mesonbuild/meson/archive/aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz -> meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz...
-- Using meson: /usr/bin/python3;/mnt/vcpkg-ci/downloads/tools/meson/meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0/meson.py
-- Performing post-build validation
-- Performing post-build validation done
```
while now:
```
Starting package 171/886: tool-meson:x64-linux
Building package tool-meson[core]:x64-linux...
-- Found meson('0.53.2') but at least version 0.58.1 is required! Trying to use internal version if possible!
-- Downloading https://github.com/mesonbuild/meson/archive/aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz -> meson-aeda7f249c4a5dbbecc52e44f382246a2377b5b0.tar.gz...
-- Using meson: /usr/bin/python3;/usr/bin/meson
-- Performing post-build validation
-- Performing post-build validation done
```

**Comment by Neumann-A** [comment]
Fix here #21341

**Pull request merged**